src/main/java/example/TriangleCount.java

```diff
@@ -0,0 +1,210 @@
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.graph.Edge;
+import org.apache.flink.graph.EdgeDirection;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.NeighborsFunctionWithVertexValue;
+import org.apache.flink.graph.Triplet;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.types.NullValue;
+import org.apache.flink.util.Collector;
+import util.TriangleCountData;
+
+import java.util.HashSet;
+import java.util.Iterator;
+
+public class TriangleCount implements ProgramDescription {
+
+	public static void main(String [] args) throws Exception {
+
+		if(!parseParameters(args)) {
+			return;
+		}
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		DataSet<Edge<String, NullValue>> edges = getEdgesDataSet(env);
+
+		Graph<String, NullValue, NullValue> graph = Graph.fromDataSet(edges, env);
+
+		// simulate the first superstep
+		// select the neighbors with id greater than the current vertex's id
+		DataSet<Vertex<String, HashSet<String>>> verticesWithLowerNeighbors =
+				graph.groupReduceOnNeighbors(new GatherLowerNeighbors(), EdgeDirection.ALL);
+
+		Graph<String, HashSet<String>, NullValue> graphWithVertexNeighbors = Graph.fromDataSet(verticesWithLowerNeighbors,
+				edges, env);
+
+		// simulate the second superstep
+		// propagate each received "message" to neighbours with higher id
+		DataSet<Vertex<String, HashSet<String>>> verticesWithPropagatedValues =
+				graphWithVertexNeighbors.groupReduceOnNeighbors(new PropagateNeighborValues(), EdgeDirection.ALL);
+
+		Graph<String, HashSet<String>, NullValue> graphWithPropagatedVertexValues = Graph.fromDataSet(verticesWithPropagatedValues,
+				edges, env);
+
+		DataSet<Vertex<String, Integer>> verticesWithNumberOfTriangles = graphWithPropagatedVertexValues.getTriplets()
+				.map(new CountTriangles());
+
+		DataSet<Tuple1<Integer>> numberOfTriangles =  verticesWithNumberOfTriangles
+				.flatMap(new FlatMapFunction<Vertex<String, Integer>, Tuple1<Integer>>() {
+
+					@Override
+					public void flatMap(Vertex<String, Integer> vertex, Collector<Tuple1<Integer>> collector) throws Exception {
+						collector.collect(new Tuple1<Integer>(vertex.getValue()));
+					}
+				}).reduce(new ReduceFunction<Tuple1<Integer>>() {
+
+					@Override
+					public Tuple1<Integer> reduce(Tuple1<Integer> firstTuple, Tuple1<Integer> secondTuple) throws Exception {
+						return new Tuple1<Integer>(firstTuple.f0 + secondTuple.f0);
+					}
+				});
+
+		// emit result
+		if(fileOutput) {
+			numberOfTriangles.writeAsCsv(outputPath, "\n", ",");
+		} else {
+			numberOfTriangles.print();
+		}
+
+		env.execute("Executing Triangle Count");
+	}
+
+	@Override
+	public String getDescription() {
+		return "Triangle Count Example";
+	}
+
+	/**
+	 * Each vertex will have the set of neighbors with id lower than its own id as value.
+	 */
+	@SuppressWarnings("serial")
+	private static final class GatherLowerNeighbors implements
```

Comment:
What this does is quite different from what I described. Here, each vertex looks at all its neighbors and chooses the ones with lower IDs. What we want is that each vertex looks at its neighbors and *generates* a msg for each neighbor with higher ID. The result is the same -- what you've implemented will give us the correct result, but if we want to closely simulate the vertex-centric algorithm, we should do it the second way. I guess a `groupReduceOnEdges` should be enough for this.